### PR TITLE
Fix router deadlock

### PR DESCRIPTION
**Pull request by no2chem**
## Overview

Description: This PR changes the behavior of NettyClientRouter so that the a created router is automatically started, and that stopping a router automatically shuts it down. This prevents a deadlock cycle that could occur due to retry holding the monitor lock on the router.

Why should this be merged: Fixes deadlock bug, needed for reconfiguration to be robust.

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 1]
```diff
@@ -325,83 +326,103 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
...
+                addReconnectionOnCloseFuture(channel, b);
+                log.info("connectAsync[{}]: Channel connected.", node);
+            } else {
+                // Otherwise, the connection failed. If we're not shutdown, try reconnecting after
+                // a sleep period.
+                if (!shutdown) {
+                    log.info("connectAsync[{}]: Channel connection failed, reconnecting...", node);
+                    Sleep.sleepUninterruptibly(parameters.getConnectionRetryRate());
+                    // Call connect, which will retry the call again.
+                    connectAsync(b);
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: This is recursive ? On node failure is it possible that we can overflow the stack ? I am not comfortable with this as this does not seem tail recursive either.

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 2]
```diff
@@ -180,35 +180,38 @@ public void setEpoch(long epoch){
...
-
-    /**
-     * Flag, if we are connected.
+    /** A {@link CompletableFuture} which is completed when a connection,
+     *  including a successful handshake completes and messages can be sent
+     *  to the remote node.
      */
     @Getter
-    volatile Boolean connected;
+    volatile CompletableFuture<Void> connectionFuture;
```
Comment (medhavidhawan) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: What is the reason for changing this ?

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 3]
```diff
@@ -63,6 +62,7 @@
 import org.corfudb.util.CFUtils;
 import org.corfudb.util.MetricsUtils;
 import org.corfudb.util.NodeLocator;
+import org.corfudb.util.NodeLocator.Protocol;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MINOR Remove this unused import 'org.corfudb.util.NodeLocator.Protocol'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 4]
```diff
@@ -12,30 +12,26 @@
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.ChannelOption;
+import io.netty.channel.DefaultChannelPromise;
 import io.netty.channel.EventLoopGroup;
 import io.netty.channel.SimpleChannelInboundHandler;
-import io.netty.channel.nio.NioEventLoopGroup;
-import io.netty.channel.socket.SocketChannel;
-import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.channel.VoidChannelPromise;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MINOR Remove this unused import 'io.netty.channel.VoidChannelPromise'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 5]
```diff
@@ -602,12 +632,28 @@ public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
...
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt.equals(ClientHandshakeEvent.CONNECTED)) {
+            // Handshake successful. Complete the connection future to allow
+            // clients to proceed.
+            connectionFuture.complete(null);
+        } else if (evt.equals(ClientHandshakeEvent.FAILED)) {
+            // Handshake failed. If the current completion future is complete,
+            // create a new one to unset it, causing future requests
+            // to wait.
+            if (connectionFuture.isDone()) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1066)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 6]
```diff
@@ -325,83 +326,103 @@ protected void initChannel(@Nonnull Channel ch) throws Exception {
...
+        }
     }
 
+    /** {@inheritDoc}
+     *  @deprecated  Deprecated, stopping a router without shutting it down is no longer supported.
+     *               Please use {@link this#stop()}.
+     */
     @Override
+    @Deprecated
     public void stop(boolean shutdown) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 7]
```diff
@@ -45,14 +41,17 @@
 import lombok.Setter;
 import lombok.extern.slf4j.Slf4j;
 
+import org.corfudb.comm.ChannelImplementation;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MINOR Remove this unused import 'org.corfudb.comm.ChannelImplementation'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 8]
```diff
@@ -271,34 +288,18 @@ public IClientRouter addClient(IClient client) {
...
-                throw new NetworkException(e.getClass().getSimpleName()
-                        + " connecting to endpoint failed", node, e);
-            }
-        }
+    @Deprecated
+    /**
+     * @deprecated The router automatically starts now, so this function call is no
+     *             longer necessary
+     */
+    public synchronized void start() {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 9]
```diff
@@ -12,30 +12,26 @@
...
-import java.net.SocketException;
+import io.netty.util.concurrent.GlobalEventExecutor;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.NoSuchElementException;
-import java.util.Random;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MINOR Remove this unused import 'java.util.concurrent.CompletionException'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [NettyClientRouter.java#L82](https://github.com/CorfuDB/CorfuDB/blob/3a3c8d582b5fdda9c2b6abc6872c968f7d28beef/runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java#L82): Remove this unused "timerConnect" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)
1. ![MAJOR][MAJOR] [NettyClientRouter.java#L84](https://github.com/CorfuDB/CorfuDB/blob/3a3c8d582b5fdda9c2b6abc6872c968f7d28beef/runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java#L84): Remove this unused "counterSendDisconnected" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java [hunk 10]
```diff
@@ -178,7 +186,7 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
      */
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx,
-                                java.lang.Throwable cause) throws Exception {
+        java.lang.Throwable cause) throws Exception {
```
[the remark this answers is not in the document]
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: Its probable that the ide did this, it's not intentional.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 11]
```diff
@@ -407,7 +407,7 @@ public CorfuRuntime registerBeforeRpcHandler(Runnable handler) {
             ? (address) -> overrideGetRouterFunction.apply(this, address) : (address) ->
                 nodeRouters.compute(address, (k, r) -> {
                     final NettyClientRouter router = (NettyClientRouter) r;
-                    if (router != null && router.getConnected()) {
+                    if (router != null) {
```
[the remark this answers is not in the document]
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: It's unclear why it was added in the first place. It seemed like an error.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 12]
```diff
@@ -422,8 +422,7 @@ public CorfuRuntime registerBeforeRpcHandler(Runnable handler) {
                                 .addClient(new SequencerClient())
                                 .addClient(new LogUnitClient().setMetricRegistry(metrics != null
                                     ? metrics : CorfuRuntime.getDefaultMetrics()))
-                                .addClient(new ManagementClient())
-                                .start();
+                                .addClient(new ManagementClient());
```
[the remark this answers is not in the document]
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: The problem of calling start multiple times. It didn't make any sense to separate them

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java (hunk 2 above)
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: Well, first, the connected variable was never properly set. Second, having a future allows a join whenever the status changes.

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java (hunk 1 above)
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: It is not recursive. The call is in a future, unlike before.

File: runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java [hunk 13]
```diff
@@ -34,6 +34,14 @@
     private final Queue<CorfuMsg> messages = new ArrayDeque();
     private static final String READ_TIMEOUT_HANDLER = "readTimeoutHandler";
 
+    /** Events that the handshaker sends to downstream handlers.
+     *
+     */
+    public enum ClientHandshakeEvent {
```
[the remark this answers is not in the document]
Comment (no2chem) [reply runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java]: This is just where the handler is, and this belongs as an inner class.

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java (hunk 1 above)
Comment (medhavidhawan) [reply runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: but connectAsync(b) is being called in connectAsync(b). if there is failure what happens ?

**Comment by no2chem** [comment]
@medhavidhawan addressed all comments

**Pull request merged**